tests/zfs-tests/include/libtest.shlib

```diff
@@ -1449,23 +1449,113 @@ function reexport_pool
 }
 
 #
-# Verify a given disk is online or offline
+# Verify a given disk or pool state
 #
 # Return 0 is pool/disk matches expected state, 1 otherwise
 #
-function check_state # pool disk state{online,offline}
+function check_state # pool disk state{online,offline,degraded}
 {
 	typeset pool=$1
 	typeset disk=${2#$DEV_DSKDIR/}
 	typeset state=$3
 
-	$ZPOOL status -v $pool | grep "$disk"  \
-	    | grep -i "$state" > /dev/null 2>&1
+	[[ -z $pool ]] || [[ -z $state ]] \
+	    && log_fail "Arguments invalid or missing"
+
+	if [[ -z $disk ]]; then
+		#check pool state only
+		$ZPOOL get -H -o value health $pool \
+		    | grep -i "$state" > /dev/null 2>&1
+	else
+		$ZPOOL status -v $pool | grep "$disk"  \
+		    | grep -i "$state" > /dev/null 2>&1
+	fi
 
 	return $?
 }
 
 #
+# Online or offline a disk on the system
+#
+# First checks state of disk. Test will fail if disk is not properly onlined
+# or offlined
+#
+function on_off_disk # disk state{online,offline} HBA CHAN TARG LUN
+{
+	typeset disk=$1
+	typeset state=$2
+	typeset HBA=$3
+	typeset CHAN=$4
+	typeset TARG=$5
+	typeset LUN=$6
+
+	[[ -z $disk ]] || [[ -z $state ]] || [[ -z $HBA ]] || \
+	    [[ -z $CHAN ]] || [[ -z $TARG ]] || [[ -z $LUN ]] \
+	    && log_fail "Arguments invalid or missing"
+
+	if is_linux; then
+		if [[ $state == "offline" ]] && ( is_mpath_device $disk ); then
+			dm_name="$($READLINK $DEV_DSKDIR/$disk \
+			    | $NAWK -F / '{print $2}')"
+			slave="$($LS /sys/block/$dm_name/slaves \
+			    | $NAWK '{print $1}')"
+			#check if disk is online
+			$LSSCSI | $EGREP $slave > /dev/null
+			if (($? == 0)); then
+				$ECHO "offline" > \
+				    /sys/class/scsi_device/${HBA}:${CHAN}:${TARG}:${LUN}/device/state
+				$ECHO "1" > \
+				    /sys/class/scsi_device/${HBA}:${CHAN}:${TARG}:${LUN}/device/delete
```

Comment:
You should be able to achieve this more realistically by offlining and removing all underlying devices from the `slaves` directory above.  This way you'll also be able to exercise the multipath code for detecting that an underlying device went away.  Plus you won't need to know the hba, chan, target, len.  Which is nice.